src/backend/distributed/commands/multi_copy.c

```diff
@@ -2145,8 +2139,15 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 	/*
 	 * Prevent concurrent UPDATE/DELETE on replication factor >1
 	 * (see AcquireExecutorMultiShardLocks() at multi_router_executor.c)
+	 *
+	 * Note that, we may need to get shard resource lock on the worker to prevent
+	 * divergence between the placement of reference table shards.
 	 */
-	LockShardListResources(shardIntervalList, RowExclusiveLock);
+	if (LockShardResourcesOnFirstWorkerIfNecessary(RowExclusiveLock, shardIntervalList) !=
+		SHARD_RESOURCES_LOCKED_LOCALLY)
+	{
+		LockShardListResources(shardIntervalList, RowExclusiveLock);
```

Comment:
This code is super confusing to me, could we maybe move the reference-table-on-mx-specific logic into `LockShardListResources` or a wrapper that does both? It could maybe also be named for its purpose, e.g. `SerialiseNonCommutativeWrites`.